### PR TITLE
.Net: Kernel as a state/property bag

### PR DESCRIPTION
**Pull request by SergeyMenshykh**
This PR is the first in a series of follow-up PRs aimed at making the SK Kernel a central component of the SK API
1. The IKernel interface is now obsolete and no longer used by the Kernel class to avoid breaking changes when adding new members. This makes it easier to extend Kernel without breaking classes that inherit from it.
2. The RunAsync method has been changed from an instance method to an extension method, as the primary purpose of the Kernel is no longer to execute functions. Now, RunAsync is just one of several operations, modeled as extension methods, that can be performed on the Kernel.

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.AI.OpenAI/OpenAIKernelBuilderExtensions.cs [hunk 1]
```diff
@@ -26,7 +26,7 @@ namespace Microsoft.SemanticKernel;
 /// <summary>
 /// Provides extension methods for the <see cref="KernelBuilder"/> class to configure OpenAI and AzureOpenAI connectors.
 /// </summary>
-public static class OpenAIKernelBuilderExtensions
+public static class OpenAKernelBuilderExtensions
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.AI.OpenAI/OpenAIKernelBuilderExtensions.cs:29]: I don't think you meant to make this change

File: dotnet/src/Connectors/Connectors.AI.OpenAI/OpenAIKernelExtensions.cs [hunk 2]
```diff
@@ -5,14 +5,14 @@
...
+// ReSharper disable once CheckNamespace - Using the namespace of Kernel
 namespace Microsoft.SemanticKernel;
 #pragma warning restore IDE0130
 
 /// <summary>
-/// Class for extension methods for <see cref="IKernel"/> using OpenAI request settings.
+/// Class for extension methods for <see cref="Kernel"/> using OpenAI request settings.
 /// </summary>
-public static class OpenAIKernelExtensions
+public static class OpenAKernelExtensions
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.AI.OpenAI/OpenAIKernelExtensions.cs:15]: Same here

**Review by markwallace-microsoft — COMMENTED** [review]
Couple of search and replace issues

File: dotnet/src/Connectors/Connectors.AI.OpenAI/OpenAIKernelBuilderExtensions.cs (hunk 1 above)
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AI.OpenAI/OpenAIKernelBuilderExtensions.cs]: Yep, right, it was not intentional change. Thanks.

File: dotnet/src/Planners/Planners.Core.UnitTests/Action/ActionPlannerTests.cs [hunk 3]
```diff
@@ -158,37 +165,39 @@ public async Task ListOfFunctionsExcludesExcludedFunctionsAsync()
...
     {
-        plugins ??= new SKPluginCollection();
-        var functionRunner = new Mock<IFunctionRunner>();
-        var serviceProvider = new Mock<IAIServiceProvider>();
-        var serviceSelector = new Mock<IAIServiceSelector>();
-        var kernel = new Mock<IKernel>();
+        if (plugins is null)
+        {
+            plugins = new SKPluginCollection();
+        }
```
Comment (stephentoub) [inline dotnet/src/Planners/Planners.Core.UnitTests/Action/ActionPlannerTests.cs:173]: This can just be:
```C#
plugin ??= new SKPluginCollection();
```

File: dotnet/src/Planners/Planners.Core.UnitTests/Sequential/SequentialPlanParserTests.cs [hunk 4]
```diff
@@ -358,4 +346,34 @@ public void CanCreatePlanWithIgnoredNodes(string goalText, string planText)
         Assert.Equal("MockPlugin", plan.Steps[1].PluginName);
         Assert.Equal("Echo", plan.Steps[1].Name);
     }
+
+    private Kernel CreateKernel(string testPlanString, SKPluginCollection? plugins = null)
+    {
+        if (plugins is null)
+        {
+            plugins = new SKPluginCollection();
+        }
```
Comment (stephentoub) [inline dotnet/src/Planners/Planners.Core.UnitTests/Sequential/SequentialPlanParserTests.cs:355]:
```suggestion
        plugins ??= new SKPluginCollection();
```

File: dotnet/src/Planners/Planners.Core.UnitTests/Sequential/SequentialPlannerTests.cs [hunk 5]
```diff
@@ -21,174 +19,122 @@ public sealed class SequentialPlannerTests
...
-    // Method to create Mock<IKernel> objects
-    private static Mock<IKernel> CreateMockKernel()
+    private Kernel CreateKernel(string testPlanString, SKPluginCollection? plugins = null)
     {
-        var kernel = new Mock<IKernel>();
-        kernel.Setup(x => x.LoggerFactory).Returns(NullLoggerFactory.Instance);
+        if (plugins is null)
+        {
+            plugins = new SKPluginCollection();
+        }
```
Comment (stephentoub) [inline dotnet/src/Planners/Planners.Core.UnitTests/Sequential/SequentialPlannerTests.cs:96]:
```suggestion
        plugins ??= new SKPluginCollection();
```

File: dotnet/src/SemanticKernel.Abstractions/IKernel.cs [hunk 6]
```diff
@@ -44,18 +43,6 @@ public interface IKernel
     [Obsolete("This will be removed in a future release. Use RegisterPlugin instead.")]
     ISKFunction RegisterCustomFunction(ISKFunction customFunction);
```
Comment (stephentoub) [inline dotnet/src/SemanticKernel.Abstractions/IKernel.cs:44]: I thought I deleted this.  Given all the breaking changes we're making, probably makes sense to delete all the Obsolete stuff that's in the repo. @markwallace-microsoft? That can obviously be a separate PR.

File: dotnet/src/SemanticKernel.Core/Kernel.cs [hunk 7]
```diff
@@ -87,6 +93,7 @@ public sealed class Kernel : IKernel, IDisposable
     /// <param name="httpHandlerFactory">HTTP handler factory</param>
     /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> to use for logging. If null, no logging will be performed.</param>
     /// <param name="serviceSelector">AI Service selector</param>
+    [Obsolete("This constructor is obsolete and will be removed in one of the next version of SK. Please use one of the available overload.")]
```
Comment (stephentoub) [inline dotnet/src/SemanticKernel.Core/Kernel.cs:96]: Related to my previous comment, should we instead just delete these?

File: dotnet/src/SemanticKernel.Core/Kernel.cs [hunk 8]
```diff
@@ -107,70 +114,53 @@ public sealed class Kernel : IKernel, IDisposable
...
+        ILoggerFactory? loggerFactory = null)
     {
-        Verify.NotNull(customFunction);
-
-        this.Plugins.Add(new SKPlugin($"plugin{Guid.NewGuid():N}", new[] { customFunction }));
+        this.LoggerFactory = loggerFactory ?? NullLoggerFactory.Instance;
+        this.HttpHandlerFactory = httpHandlerFactory ?? NullHttpHandlerFactory.Instance;
+        this._aiServiceProvider = aiServiceProvider;
+        this.Plugins = plugins ?? new SKPluginCollection();
+        this._aiServiceSelector = serviceSelector ?? new OrderedIAIServiceSelector();
```
Comment (stephentoub) [inline dotnet/src/SemanticKernel.Core/Kernel.cs:136]: We could make these cheaper by creating them lazily; that way they won't be instantiated unless used.  But if we expect not using them to be rare, it's fine and simpler to create them here in the ctor.

**Review by stephentoub — APPROVED** [review]

File: dotnet/src/SemanticKernel.Abstractions/IKernel.cs (hunk 6 above)
Comment (markwallace-microsoft) [reply dotnet/src/SemanticKernel.Abstractions/IKernel.cs:44]: Agreed, let's clean out obsolete stuff as we go

**Review by markwallace-microsoft — APPROVED** [review]

File: dotnet/src/Planners/Planners.Core.UnitTests/Action/ActionPlannerTests.cs (hunk 3 above)
Comment (SergeyMenshykh) [reply dotnet/src/Planners/Planners.Core.UnitTests/Action/ActionPlannerTests.cs]: Yep, fixed.

File: dotnet/src/SemanticKernel.Core/Kernel.cs (hunk 8 above)
Comment (SergeyMenshykh) [reply dotnet/src/SemanticKernel.Core/Kernel.cs:136]: A good idea that I would suggest deferring for a later stage after we have completed cloning.
